### PR TITLE
osd: Try other PGs when reservation failures occur

### PR DESCRIPTION
**Pull request by dzafman**
## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by dzafman** [comment]
I removed the local reservation pre-check.  This makes the extremely rare "local reserve failed" a normal scenario.  These log lines shows that things are working as expected.

```
2021-01-12T10:11:25.925-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.8 local reserve failed, nothing to be done now

2021-01-12T10:11:28.949-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.8 reserve failed, skipped
2021-01-12T10:11:28.949-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.10 local reserve failed, nothing to be done now

2021-01-12T10:11:30.909-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.8 reserve failed, skipped
2021-01-12T10:11:30.909-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.10 reserve failed, skipped

2021-01-12T10:11:32.898-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.8 reserve failed, skipped
2021-01-12T10:11:32.898-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.10 reserve failed, skipped
2021-01-12T10:11:32.898-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.16 reserve failed, skipped

2021-01-12T10:11:34.890-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.8 reserve failed, skipped
2021-01-12T10:11:34.890-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.10 reserve failed, skipped
2021-01-12T10:11:34.890-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.16 reserve failed, skipped
2021-01-12T10:11:34.890-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.18 reserve failed, skipped

2021-01-12T10:11:37.946-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.8 reserve failed, skipped
2021-01-12T10:11:37.946-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.10 reserve failed, skipped
2021-01-12T10:11:37.946-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.16 reserve failed, skipped
2021-01-12T10:11:37.946-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.18 reserve failed, skipped

2021-01-12T10:11:39.938-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.8 local reserve failed, nothing to be done now

2021-01-12T10:11:42.006-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.8 reserve failed, skipped
2021-01-12T10:11:42.006-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.10 local reserve failed, nothing to be done now

2021-01-12T10:11:43.042-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.8 reserve failed, skipped
2021-01-12T10:11:43.042-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.10 reserve failed, skipped
2021-01-12T10:11:43.042-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.16 local reserve failed, nothing to be done now

2021-01-12T10:11:45.034-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.8 reserve failed, skipped
2021-01-12T10:11:45.034-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.10 reserve failed, skipped
2021-01-12T10:11:45.034-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.16 reserve failed, skipped

2021-01-12T10:11:46.042-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.8 reserve failed, skipped
2021-01-12T10:11:46.042-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.10 reserve failed, skipped
2021-01-12T10:11:46.042-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.16 reserve failed, skipped
2021-01-12T10:11:46.042-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.18 reserve failed, skipped

2021-01-12T10:11:50.046-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.10 local reserve failed, nothing to be done now

2021-01-12T10:11:51.034-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.10 reserve failed, skipped
2021-01-12T10:11:51.034-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.16 local reserve failed, nothing to be done now

2021-01-12T10:11:52.070-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.10 reserve failed, skipped
2021-01-12T10:11:52.070-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.16 reserve failed, skipped
2021-01-12T10:11:52.070-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.18 local reserve failed, nothing to be done now

2021-01-12T10:11:54.078-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.10 reserve failed, skipped
2021-01-12T10:11:54.078-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.16 reserve failed, skipped
2021-01-12T10:11:54.078-0800 7f52d1efd700 20 osd.5 37 sched_scrub pg  1.18 reserve failed, skipped
```

File: src/osd/OSD.cc [hunk 1]
```diff
@@ -7630,7 +7637,7 @@ void OSD::sched_scrub()
       if (pg->m_scrubber->is_reserving()) {
 	pg->unlock();
 	dout(10) << __func__ << ": reserve in progress pgid " << scrub_job.pgid << dendl;
-	break;
+	goto out;
```
Comment (ronen-fr) [inline src/osd/OSD.cc:7640]: why 'goto' instead of the 'break'?

File: src/osd/PG.h [hunk 2]
```diff
@@ -186,6 +186,10 @@ class PG : public DoutPrefixProvider, public PeeringState::PeeringListener {
   /// scrubbing state for both Primary & replicas
   bool is_scrub_active() const { return m_scrubber->is_scrub_active(); }
 
+  bool get_reserve_failed() const { return m_scrubber->get_reserve_failed(); }
```
Comment (ronen-fr) [inline src/osd/PG.h:189]: I would prefer, if we choose to have this status flag, to have it as part of the object that actually holds that status -
m_local_osd_resource.

File: src/osd/pg_scrubber.h [hunk 3]
```diff
@@ -413,6 +413,10 @@ class PgScrubber : public ScrubPgIF, public ScrubMachineListener {
 
   [[nodiscard]] bool is_scrub_active() const final { return m_active; }
 
+  [[nodiscard]] bool get_reserve_failed() const final { return m_reserve_failed; }
```
Comment (ronen-fr) [inline src/osd/pg_scrubber.h:416]: Should be part of the m_local_osd_resource object.

Seems to me that the clearing API shouldn't exist.

**Review by ronen-fr — COMMENTED** [review]
I would recommend not spreading the "local resource locking" handling outside of
PgScrubber::m_local_osd_resource

File: src/osd/PG.h (hunk 2 above)
Comment (dzafman) [reply src/osd/PG.h:189]: What do you mean by object?  This is really the PG that is marked.

File: src/osd/pg_scrubber.h (hunk 3 above)
Comment (dzafman) [reply src/osd/pg_scrubber.h:416]: This flag represents a temporary state of the PG.  When other scrubs finish, reservations can succeed. This flag is a way to cycle through all possible scrubs to try to find one that can run now.  We have to keep trying in order, so after 1 cycle we clear all flags to start over.

File: src/osd/PG.h (hunk 2 above)
Comment (ronen-fr) [reply src/osd/PG.h:189]: The status of "being locked for scrubbing" is now managed - indeed inside the PG, but within sub-elements of the
PgScrubber. It shouldn't be spread out.

**Comment by ronen-fr** [comment]
Not sure I understand the idea: if to mark PGs that failed to reserve their replicas - good. And calling
set_reserve_failed on scrub_machine.cc:99 makes sense. But then why set it also when failing to reserve the local
OSD (PG.cc:1364) ? if the local OSD is overwhelmed, no point in trying the other PGs we manage.

File: src/osd/OSD.cc (hunk 1 above)
Comment (dzafman) [reply src/osd/OSD.cc:7640]: We don't _break_ here because I don't want to clear the reserve_failed flags from all PGs.

File: src/osd/PG.h (hunk 2 above)
Comment (dzafman) [reply src/osd/PG.h:189]: What does "spread out" mean?

**Comment by dzafman** [comment]
> Not sure I understand the idea: if to mark PGs that failed to reserve their replicas - good. And calling
> set_reserve_failed on scrub_machine.cc:99 makes sense. But then why set it also when failing to reserve the local
> OSD (PG.cc:1364) ? if the local OSD is overwhelmed, no point in trying the other PGs we manage.

Most of the time if local reservation aren't available then osd/OSD.cc:7569 pre-check (doesn't grab any reservations) will fast out of OSD::sched_scrub().  In a very rare race where the last reservation is grabbed by a remote scrub before we try to get a local reservation then setting reserve_failed set in PG.cc:1364 is detached at osd/OSD.cc:7652 and we can just jump out of the loop.

File: src/osd/PG.h (hunk 2 above)
Comment (ronen-fr) [reply src/osd/PG.h:189]: What I meant is this: shouldn't the status of 'I have failed trying to lock my replicas' be managed by the same
object that manages the status 'what replicas I have tried to have resource-locked for me, and what's the status with
that locking?' (ReplicaReservations)

(I was wrong when ref to m_local_osd_resource (see my question on that below))

**Comment by ronen-fr** [comment]
> > Not sure I understand the idea: if to mark PGs that failed to reserve their replicas - good. And calling
> > set_reserve_failed on scrub_machine.cc:99 makes sense. But then why set it also when failing to reserve the local
> > OSD (PG.cc:1364) ? if the local OSD is overwhelmed, no point in trying the other PGs we manage.
> 
> Most of the time if local reservation aren't available then osd/OSD.cc:7569 pre-check (doesn't grab any reservations) will fast out of OSD::sched_scrub(). In a very rare race where the last reservation is grabbed by a remote scrub before we try to get a local reservation then reserve_failed set in PG.cc:1364 is detected at osd/OSD.cc:7652 and we can just jump out of the loop.

I still didn't get what we gain by setting the flag on local reservation failure, but will be happy to take this discussion offline.

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-01-14_16:22:32-rados-wip-kefu-testing-2021-01-14-2241-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-01-15_15:58:16-rados-wip-kefu-testing-2021-01-15-1726-distro-basic-smithi/

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by dzafman** [comment]
jenkins test make check

File: src/osd/OSD.cc [hunk 4]
```diff
@@ -7639,11 +7646,34 @@ void OSD::sched_scrub()
       if (pg->sched_scrub()) {
 	pg->unlock();
         dout(10) << __func__ << " scheduled a scrub!" << " (~" << scrub_job.pgid << "~)" << dendl;
-	break;
+	goto out;
       }
+      // If this is set now we must have had a local reserve failure, so can't scrub anything right now
+      if (pg->get_reserve_failed()) {
```
Comment (dzafman) [inline src/osd/OSD.cc:7652]: @ronen-fr This check is why I mark a PG for local reservation failure too.

**Review by jdurgin — APPROVED** [review]

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-02-02_15:01:12-rados-wip-kefu-testing-2021-02-02-1939-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-02-03_08:10:00-rados-wip-kefu-testing-2021-02-03-1357-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-02-03_14:17:43-rados-wip-kefu-testing-2021-02-03-2051-distro-basic-smithi/

**Comment by neha-ojha** [comment]
jenkins test make check

**Review by neha-ojha — APPROVED** [review]

**Pull request merged**